### PR TITLE
CA-500 Improve Bond error codes and logging

### PR DESCRIPTION
**Pull request by wchamber**
Change Bond's error for unable to find refresh tokens to 404 instead of 400. Suggest to users that they relink their accounts to Bond.
Add some logging statements in odder corner cases of the code.

Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/bond/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've followed [the instructions](https://github.com/DataBiosphere/bond/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [ ] I've updated the [FISMA documentation](https://github.com/DataBiosphere/bond/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary. Documentation PRs only need 1 thumb.
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this **unless it's for a hotfix**. In that case, don't delete it!
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

**Comment by wchamber** [comment]
I'm opening this for review now as merging into `python3-feature`, but I think we should wait to merge this into develop until we've already successfully deployed py3 Bond.
I looked at all of the places that we're raising exceptions in Bond.

File: bond_app/fence_token_vending.py [hunk 1]
```diff
@@ -62,15 +62,9 @@ def get_service_account_key_json(self, user_info):
...
-        real_user_info = self.sam_api.user_info(user_info.token)
-        if real_user_info is None:
-            raise exceptions.Unauthorized("user not found in sam")
-        return real_user_info
-
     def _get_oauth_access_token(self, provider_user):
         refresh_token = self.refresh_token_store.lookup(provider_user.user_id, self.provider_name)
         if refresh_token is None:
-            raise exceptions.BadRequest("Fence account not linked")
+            raise exceptions.NotFound("Fence account not linked. {}".format(str(provider_user)))
```
Comment (dvoet) [inline bond_app/fence_token_vending.py:73]: What is the rationale for this change? I don't recall what we were thinking at the time but BadRequest must have made sense.

File: bond_app/sam_api.py [hunk 2]
```diff
@@ -12,14 +12,14 @@ def user_info(self, access_token):
...
-        :return: dict with userSubjectId and userEmail keys or None if user does not exist in sam
+        :return: dict with userSubjectId and userEmail keys or else raises an exception if user does not exist in sam
         """
         headers = {'Authorization': 'Bearer ' + access_token}
         result = requests.get(url=self.base_url + '/register/user/v1?userDetailsOnly=true', headers=headers)
         if result.status_code == 200:
             return json.loads(result.content)["userInfo"]
         elif result.status_code == 404:
-            return None
+            raise exceptions.Unauthorized("user not found in sam")
```
Comment (dvoet) [inline bond_app/sam_api.py]: I disagree with this change. The sam_api module should reflect the sam api and let the caller determine what to do with the response. The appropriate return value for a 404 is None and the caller should raise the appropriate exception or handle it

**Review by dvoet — CHANGES_REQUESTED** [review]

File: bond_app/sam_api.py (hunk 2 above)
Comment (gpolumbo-broad) [reply bond_app/sam_api.py]: That being said, if the status is anything other than `2xx` here, we should add a logging statement to log the response from Sam.

**Review by gpolumbo-broad — APPROVED** [review]

File: bond_app/sam_api.py (hunk 2 above)
Comment (wchamber) [reply bond_app/sam_api.py]: sg, reverted fence_token_vending and added checks to bond.py to handle None return case.
Also added logging here.

File: bond_app/fence_token_vending.py (hunk 1 above)
Comment (wchamber) [reply bond_app/fence_token_vending.py:73]: I don't think BadRequest is a good match for what's going on. Bond wasn't apple to find a valid refresh token for the user, either because it expired or the user has never linked their account. There's nothing bad about the user's request. They wouldn't have to change the request if their token existed, so I don't think BadRequest is a good fit.

I'm not sure if Unauthorized, Forbidden, or NotFound is best. We didn't find the refresh token, we're not sure why. The user might just need to authorize their fence account through Bond, which sounds Unauthorized-ish to me. Or maybe the current action is Forbidden since we know who the user is in Sam, but they don't have a refresh token that matches. WDYT?

**Review by gpolumbo-broad — APPROVED** [review]

**Review by dvoet — APPROVED** [review]
